File: typescript/src/workshop/m7/s12_returnasap/step1/SeatFinder.ts
import type { Seat } from '../Seat.js';

/**
 * Krok 1: guard clauses - w seatClass() przypadki kończące (brak, zajęte) zwracają od razu,
 * zmienna result zniknęła. W firstFree() guard dla null zamiast otaczającego if.
 */
export class SeatFinder {
  private inspectedCount = 0;

  seatClass(seat: Seat | null, vipFromRow: number): string {
    if (seat === null) {
      return 'BRAK';
    }
    if (seat.taken) {
      return 'ZAJETE';
    }
    if (seat.row >= vipFromRow) {
      return 'VIP';
    }
    return 'STANDARD';
  }

  firstFree(seats: readonly Seat[] | null, minRow: number): string | undefined {
    if (seats === null) {
      return undefined;
    }
    let result: string | undefined = undefined;
    let found = false;
    let index = 0;
    while (!found && index < seats.length) {
      const seat = seats[index]!;
      this.inspectedCount++;
      if (seat.row >= minRow && !seat.taken) {
        result = seat.label;
        found = true;
      }
      index++;
    }
    return result;
  }

  inspected(): number {
    return this.inspectedCount;
  }
}

File: csharp/src/Training.Workshop/M7/S12ReturnAsap/Step2/SeatFinder.cs
namespace Training.Workshop.M7.S12ReturnAsap.Step2;

/// <summary>
/// Krok 2 (rozwiązanie): Return ASAP w pętli - zwracamy tam, gdzie wynik jest ostateczny.
/// Flaga found i zmienna result zniknęły. _inspected++ zostaje PRZED return (mutacja zachowana),
/// a Count/seats[index] zostają - foreach (enumerator) zmieniłby sposób dostępu do listy.
/// </summary>
public sealed class SeatFinder
{
    private int _inspected;

    public string SeatClass(Seat? seat, int vipFromRow)
    {
        if (seat == null)
        {
            return "BRAK";
        }
        if (seat.Taken)
        {
            return "ZAJETE";
        }
        if (seat.Row >= vipFromRow)
        {
            return "VIP";
        }
        return "STANDARD";
    }

    public string? FirstFree(IReadOnlyList<Seat>? seats, int minRow)
    {
        if (seats == null)
        {
            return null;
        }
        for (var index = 0; index < seats.Count; index++)
        {
            var seat = seats[index];
            _inspected++;
            if (seat.Row >= minRow && !seat.Taken)
            {
                return seat.Label;
            }
        }
        return null;
    }

    public int Inspected => _inspected;
}
